module/zfs/zio.c

```diff
@@ -1626,15 +1615,25 @@ zio_vdev_delegated_io(vdev_t *vd, uint64_t offset, abd_t *data, uint64_t size,
 	return (zio);
 }
 
+
+/*
+ * Send a flush command to the given vdev. Unlike most zio creation functions,
+ * the flush zios are issued immediately. You can wait on pio to pause until
+ * the flushes complete.
+ */
 void
 zio_flush(zio_t *pio, vdev_t *vd)
 {
+	const zio_flag_t flags = ZIO_FLAG_CANFAIL | ZIO_FLAG_DONT_PROPAGATE |
+	    ZIO_FLAG_DONT_RETRY;
+
 	if (vd->vdev_nowritecache)
 		return;
+
 	if (vd->vdev_children == 0) {
-		zio_nowait(zio_ioctl(pio, vd->vdev_spa, vd,
-		    DKIOCFLUSHWRITECACHE, NULL, NULL, ZIO_FLAG_CANFAIL |
-		    ZIO_FLAG_DONT_PROPAGATE | ZIO_FLAG_DONT_RETRY));
+		zio_nowait(zio_create(pio, pio->io_spa, 0, NULL, NULL, 0, 0,
```

Comment:
Thanks. But still, why pio->io_spa, not vd->vdev_spa if vd is a mandatory argument any way?

Reply:
I don't think I have a reason, I think I just wrote it that way. If anything, its because this patch came out of a much larger unit of work on a customer branch, based on 2.1.5. Just means I should take more care with my comparison next time!

I'm not sure exactly what point you're making though. I guess this way it means that it can be called as `zio_flush(NULL, vd)`, if you wanted to for whatever reason? Regardless, it falls out the same way for me, so I've changed it.